Reverse summation order in nested_kl_divergence() in PPO

v-sivak opened the pull request:
nested_kl_divergence() in ppo_utils.py wasn't summing up the individual contributions to KL correctly, if the nested actions have different shapes. In this commit the order of operations is reversed: first the sum is done over time steps, which brings all KL contributions to the same shape, and then the sum over actions produces a total_kl. Fixes #439.

v-sivak committed:
Fix nested_kl_divergence() for nested actions

Fixes #439.

v-sivak commented:
@googlebot I signed it!

summer-yue commented on tf_agents/agents/ppo/ppo_utils.py:
```diff
@@ -126,14 +126,17 @@ def nested_kl_divergence(nested_from_distribution: types.NestedDistribution,
-    reduce_dims.remove(dim)
-  total_kl = tf.reduce_sum(input_tensor=summed_kl_divergences, axis=reduce_dims)
+  total_kl = tf.add_n(all_kl_divergences_reduced)
 
```
Great find, and thanks for the contribution! Could you please add a unit test for this to make sure it works? Can probably copy most of the existing `test_nested_kl_divergence` and change the action spec.

v-sivak replied:
Thanks! I agree with your bug philosophy. Added a unit test.

tfboyd commented:
Sorry for the delay.  I was out. I will try to nudge someone into taking a look on Monday. We **were** doing so well on reviewing PRs.

ebrevdo commented:
@summer-yue ptal?

summer-yue reviewed (approved):
This PR should fix the bug which assumed the shapes of components in the action to be the same. There's probably faster ways to do this but no bug is better than bug :) Thanks for the contribution!

v-sivak committed:
Add unit test for nested_kl_divergence in PPO

This test makes sure that nested_kl_divergence works with nested distributions of different shapes.

v-sivak commented:
By the way, unrelated to this PR, but I've written a `FlattenObservationWrapper` for tf policies, because a similar wrapper from `environments/wrappers` only works with py policies. I could push that too if you think it's useful (currently your `environments/tf_wrappers` is almost empty).

tfboyd commented:
@v-sivak  Trying to remember to ask the team today about your FlattenObservationWrapper and get back to you after.

tfboyd commented:
@ebrevdo  Will give you some info on the  `FlattenObservationWrapper` we had a great conversation and want to share the direction with you.

v-sivak commented:
Hey @ebrevdo any news here?

ebrevdo commented:
Apologies for not responding sooner.  Summer's approval LGTM; we should merge this.

Separately about `FlattenObserverWrapper`, we should try to avoid making envs that lose information by flattening.  Instead, we are working on mmore flexible network architectures that will allow you to process non-flattened data more easily.  A good example is [here](https://github.com/tensorflow/agents/blob/367abe3596d3efde92857d43f8d9642895383e1b/tf_agents/networks/nest_map_test.py#L82).  In particular, we've added a new `Sequential`, `NestMap`, and `NestFlatten` networks, which allow you to combine, build towers and flatten early on in your Network.

We hope this new approach to building networks can replace the old monolithic networks (e.g. QNetwork for DQN) which try to do everything in one place and aren't very flexible.

We're working on extending support for this to our actor networks, which have the additional complexity that they must emit `tfp.Distribution` objects.  In the meantime, you can use the new `Sequential` network for DQN, Behavioral Cloning, and any critic network.
